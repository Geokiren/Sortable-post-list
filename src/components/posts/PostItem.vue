<template>
  <div class="post-item card p-2 md:p-4 h-24 lg:h-32 mb-4 md:mb-6 flex justify-between items-center text-text-primary text-xl font-semibold" data-test="post-item">
    <div>Post {{ id }}</div>

    <div class="flex flex-col h-full" :class="buttonsContainerClass">
      <button v-if="showUp" class="caret" @click="getUp">
        <CaretSvg classes="w-5 h-5 fill-secondary" />
      </button>

      <button v-if="showDown" class="caret rotate-180" @click="getDown">
        <CaretSvg classes="w-5 h-5 fill-secondary" />
      </button>
    </div>
  </div>
</template>

<script setup lang="ts">
  import { computed } from 'vue';
  import CaretSvg from '../shared/CaretSvg.vue';
  import { usePostsStore } from '@/stores/posts';

  const props = defineProps<{
    id: number,
    index: number,
    size: number,
  }>();

  const { swapIndexes }: { swapIndexes: (from: number, to: number) => void } = usePostsStore();

  const getDown = () => {
    if (typeof props.index === 'number') {
      swapIndexes(props.index, props.index + 1);
    }
  }

  const getUp = () => {
    if (typeof props.index === 'number') {
      swapIndexes(props.index, props.index - 1);
    }
  }

  const showUp = computed(() => typeof props.index === 'number' && props.index > 0);
  const showDown = computed(() => typeof props.index === 'number' && props.index < props.size - 1);
  const buttonsContainerClass = computed(() => (showUp.value && showDown.value) ? 'justify-between' : 'justify-center');
</script>