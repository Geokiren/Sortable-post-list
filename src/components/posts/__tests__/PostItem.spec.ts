import { describe, it, expect, beforeEach } from 'vitest';
import { mount } from '@vue/test-utils';
import PostItem from '../PostItem.vue';
import CaretSvg from '@/components/shared/CaretSvg.vue';
import { createPinia, setActivePinia } from 'pinia';

describe('PostItem', () => {
  beforeEach(() => {
    setActivePinia(createPinia());
  });

  it('Renders properly with the correct content', () => {
    const wrapper = mount(PostItem, { props: { id: 1, index: 1, size: 5 } });

    const tag = wrapper.element.tagName.toLowerCase();
    const caret = wrapper.findComponent(CaretSvg);

    expect(tag).toBe('div');
    expect(wrapper.text()).toBe('Post 1');
    expect(caret.exists()).toBe(true);
  });

  it('Renders properly with one carret element', () => {
    const wrapper = mount(PostItem, { props: { id: 1, index: 0, size: 5 } });

    const carets = wrapper.findAll('[data-test="caret-svg"]');

    expect(carets.length).toBe(1);
  });

  it('Renders properly with two carret elements', () => {
    const wrapper = mount(PostItem, { props: { id: 1, index: 1, size: 5 } });

    const carets = wrapper.findAll('[data-test="caret-svg"]');

    expect(carets.length).toBe(2);
  });

  it('Matches snapshot', () => {
    const wrapper = mount(PostItem, { props: { id: 1, index: 0, size: 5 } });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
