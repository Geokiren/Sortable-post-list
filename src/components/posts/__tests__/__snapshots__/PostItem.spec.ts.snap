// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PostItem > Matches snapshot 1`] = `
"<div class="post-item card p-2 md:p-4 h-24 lg:h-32 mb-4 md:mb-6 flex justify-between items-center text-text-primary text-xl font-semibold" data-test="post-item">
  <div>Post 1</div>
  <div class="flex flex-col h-full justify-center">
    <!--v-if--><button class="caret rotate-180"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512" class="w-5 h-5 fill-secondary" data-test="caret-svg">
        <path d="M201.4 137.4c12.5-12.5 32.8-12.5 45.3 0l160 160c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L224 205.3 86.6 342.6c-12.5 12.5-32.8 12.5-45.3 0s-12.5-32.8 0-45.3l160-160z"></path>
      </svg></button>
  </div>
</div>"
`;
